File: python/611_有效三角形的个数_medium.py
class Solution:
    def triangleNumber(self, nums: List[int]) -> int:
        n = len(nums)
        nums.sort()
        ans = 0
        for i in range(0, n-2):
            k = i+1
            for j in range(i+1, n-1):
                while k+1 < n and nums[k+1] < nums[i] + nums[j]:
                    k += 1
                ans += max(k-j, 0)
        return ans

File: cpp/611_有效三角形的个数_medium.cpp
class Solution {
public:
    int triangleNumber(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        int ans = 0;
        for (int i=0; i<n-2; i++) {
            int k = i+1;
            for (int j=i+1; j<n-1; j++) {
                while (k+1 < n && nums[k+1] < nums[i] + nums[j])
                    ++k;
                ans += max(k-j, 0);
            }
        }
        return ans;
    }
};